db/compaction/compaction_iterator.cc

```diff
@@ -221,39 +221,43 @@ void CompactionIterator::Next() {
 
 bool CompactionIterator::InvokeFilterIfNeeded(bool* need_skip,
                                               Slice* skip_until) {
+  if (!compaction_filter_) {
+    return true;
+  }
+
   // TODO: support compaction filter for wide-column entities
-  if (!compaction_filter_ ||
-      (ikey_.type != kTypeValue && ikey_.type != kTypeBlobIndex)) {
+  if (ikey_.type != kTypeValue && ikey_.type != kTypeBlobIndex) {
     return true;
   }
-  bool error = false;
-  // If the user has specified a compaction filter and the sequence
-  // number is greater than any external snapshot, then invoke the
-  // filter. If the return value of the compaction filter is true,
-  // replace the entry with a deletion marker.
-  CompactionFilter::Decision filter = CompactionFilter::Decision::kUndetermined;
-  compaction_filter_value_.clear();
-  compaction_filter_skip_until_.Clear();
+
+  CompactionFilter::Decision decision =
+      CompactionFilter::Decision::kUndetermined;
   CompactionFilter::ValueType value_type =
       ikey_.type == kTypeValue ? CompactionFilter::ValueType::kValue
                                : CompactionFilter::ValueType::kBlobIndex;
+
   // Hack: pass internal key to BlobIndexCompactionFilter since it needs
   // to get sequence number.
   assert(compaction_filter_);
-  Slice& filter_key =
-      (ikey_.type == kTypeValue ||
+  const Slice& filter_key =
+      (ikey_.type != kTypeBlobIndex ||
```

Comment:
Is this intentional to check the type from kTypeValue to all types except kTypeBlobIndex?

Reply:
Yes, it's intentional; we only have to pass the internal key for blob indexes to the old BlobDB's special compaction filter (in all other cases, we have to pass the user key).

Right now, this code path is only taken for `kTypeValue` and `kTypeBlobIndex`, so the old and new conditions are actually equivalent; however, compaction filter support for wide-column entities is coming soon and for `kTypeWideColumnEntity` we would also want to pass the user key as usual.